Archives/VolFile.cpp

```diff
@@ -298,50 +285,30 @@ namespace Archives
 		volWriter.Write(&padding, volInfo.paddedIndexTableLength - volInfo.indexTableLength);
 	}
 
-	bool VolFile::OpenAllInputFiles(CreateVolumeInfo &volInfo)
+	void VolFile::OpenAllInputFiles(CreateVolumeInfo &volInfo)
 	{
-		volInfo.stringTableLength = 0;
-
-		// Open all the input files and store indexing info
-		for (std::size_t i = 0; i < volInfo.fileCount(); i++)
-		{
-			HANDLE fileHandle = CreateFileA(
-				volInfo.filesToPack[i].c_str(),	// fileName
-				GENERIC_READ,			// access mode
-				0,						// share mode
-				nullptr,				// security attributes
-				OPEN_EXISTING,			// creation disposition
-				FILE_ATTRIBUTE_NORMAL,	// file attributes
-				nullptr);				// template
-			
-			volInfo.fileHandles.push_back(fileHandle);
-
-			if (fileHandle == INVALID_HANDLE_VALUE)// Check for errors
-			{
-				// Error opening file. Close already opened files and return error
-				for (i--; i >= 0; i--) {
-					CloseHandle(volInfo.fileHandles[i]);
-				}
-
-				return false;
+		for (const auto& filename : volInfo.filesToPack) {
+			try {
+				volInfo.fileStreamReaders.push_back(std::make_unique<FileStreamReader>(filename));
+			}
+			catch (const std::exception& e) {
+				throw std::runtime_error("Error attempting to open " + filename + 
+					" for reading into volume " + m_ArchiveFileName + ". Internal Error: " + e.what());
 			}
 		}
-
-		return true;
 	}
 
-	bool VolFile::PrepareHeader(CreateVolumeInfo &volInfo)
+	void VolFile::PrepareHeader(CreateVolumeInfo &volInfo)
 	{
-		if (!OpenAllInputFiles(volInfo)) {
-			return false;
-		}
+		OpenAllInputFiles(volInfo);
 
 		// Get file sizes and calculate length of string table
 		for (std::size_t i = 0; i < volInfo.fileCount(); i++)
 		{
 			IndexEntry indexEntry;
 
-			indexEntry.fileSize = GetFileSize(volInfo.fileHandles[i], 0);
+			// A packed volume cannot be larger than uint32_t
+			indexEntry.fileSize = static_cast<uint32_t>(volInfo.fileStreamReaders[i]->Length());
```

Comment:
We should probably check this and throw an exception if it's not true. There's no way to know for certain that a file won't exceed 4GB in size until run time.

Reply:
Good call. We should look at mirroring this in clmFile after we replace all the stream related variables to fixed width in clmFile.